src/ferui-components/tree-view/tree-view-component.ts

```diff
@@ -0,0 +1,382 @@
+   */
+  private rebuildVirtualScrollerArray(): void {
+    const aggregator = [];
+    if (this.rootNode) {
+      this.addNodeAndChildrenToVirtualScrollerArray(this.rootNode, aggregator);
+    } else {
+      for (const child of this.nonRootArray) {
+        this.addNodeAndChildrenToVirtualScrollerArray(child, aggregator);
+      }
+    }
+    this.scrollViewArray = aggregator;
+    this.cd.markForCheck();
+  }
+
+  /**
+   * Returns whether the child node branch is complete and all its children loaded
+   * @param {TreeNode<T>} node
+   * @param {Array<TreeNode<T>>} aggregator
+   * @returns {boolean}
+   */
+  private addNodeAndChildrenToVirtualScrollerArray(node: TreeNode<T>, aggregator: Array<TreeNode<T>>): boolean {
+    aggregator.push(node);
+    for (const child of node.children) {
+      if (!this.addNodeAndChildrenToVirtualScrollerArray(child, aggregator)) {
+        return false;
+      }
+    }
+    return node.expanded === false || node.allChildrenLoaded;
+  }
+
+  /**
+   * Gets the first child node with more children to load
+   * @param {TreeNode<T>} node
+   * @returns {TreeNode<T>}
+   */
+  private getFirstNodeWithMoreChildrenToLoad(node: TreeNode<T>): TreeNode<T> | null {
+    if (node.expanded === true && !node.allChildrenLoaded) {
+      return node;
+    }
+    for (const child of node.children) {
+      const childResult = this.getFirstNodeWithMoreChildrenToLoad(child);
+      if (childResult !== null) {
+        return childResult;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Handles the expansion of a Tree Node component
+   * @param {TreeNode<T>} node
+   * @param {number} firstIdxInView
+   * @param {number} lastIdxInView
+   * @returns {Promise<void>}
+   */
+  private async handleExpandNode(node: TreeNode<T>, firstIdxInView: number, lastIdxInView: number): Promise<void> {
+    node.showLoader = node.expanded = true;
+    await this.loadMoreNodes(node, this.bufferAmount + lastIdxInView - firstIdxInView, false);
+    node.showLoader = false;
+    this.rebuildVirtualScrollerArray();
+  }
+
+  /**
+   * Handles the collapse event of a Tree Node Component
+   * @param {TreeNode<T>} node
+   * @returns {Promise<void>}
+   */
+  private async handleCollapseNode(node: TreeNode<T>): Promise<void> {
+    node.children = [];
+    node.expanded = node.showLoader = node.loadError = false;
+    node.allChildrenLoaded = false;
+    this.rebuildVirtualScrollerArray();
+  }
+
+  /**
+   * Handles the scroll event of the Tree View
+   * @param {number} lastIdxInView
+   * @returns {Promise<void>}
+   */
+  private async handleScroll(lastIdxInView: number): Promise<void> {
+    const numberNeeded = this.bufferAmount - (this.scrollViewArray.length - lastIdxInView);
+    if (numberNeeded > 0 && !this.scrollPromise) {
+      this.scrollPromise = true;
+      await this.loadMoreNodes(this.scrollViewArray[lastIdxInView].parent, numberNeeded, true);
+      this.rebuildVirtualScrollerArray();
+    }
+    this.cd.markForCheck();
```

Comment:
This should be called only if your condition is true because you're not updating any variables otherwise.